// joi/lib/index.js
// Load modules

var Hoek = require('hoek');
var Any = require('joi/lib/any');
var Cast = require('joi/lib/cast');
var Ref = require('joi/lib/ref');


// Declare internals

var internals = {
    alternatives: require('joi/lib/alternatives'),
    array: require('joi/lib/array'),
    boolean: require('joi/lib/boolean'),
    binary: require('joi/lib/binary'),
    date: require('joi/lib/date'),
    func: require('joi/lib/function'),
    number: require('joi/lib/number'),
    object: require('joi/lib/object'),
    string: require('joi/lib/string')
};


internals.root = function () {

    var any = new Any();

    var root = any.clone();
    root.any = function () {

        return any;
    };

    root.alternatives = root.alt = function () {

        return arguments.length ? internals.alternatives.try.apply(internals.alternatives, arguments) : internals.alternatives;
    };

    root.array = function () {

        return internals.array;
    };

    root.boolean = root.bool = function () {

        return internals.boolean;
    };

    root.binary = function () {

        return internals.binary;
    };

    root.date = function () {

        return internals.date;
    };

    root.func = function () {

        return internals.func;
    };

    root.number = function () {

        return internals.number;
    };

    root.object = function () {

        return arguments.length ? internals.object.keys.apply(internals.object, arguments) : internals.object;
    };

    root.string = function () {

        return internals.string;
    };

    root.ref = function () {

        return Ref.create.apply(null, arguments);
    };

    root.isRef = function (ref) {

        return Ref.isRef(ref);
    };

    root.validate = function (value /*, [schema], [options], callback */) {

        var last = arguments[arguments.length - 1];
        var callback = typeof last === 'function' ? last : null;

        var count = arguments.length - (callback ? 1 : 0);
        if (count === 1) {
            return any.validate(value, callback);
        }

        var options = count === 3 ? arguments[2] : {};
        var schema = Cast.schema(arguments[1]);

        return schema._validateWithOptions(value, options, callback);
    };

    root.describe = function () {

        var schema = arguments.length ? Cast.schema(arguments[0]) : any;
        return schema.describe();
    };

    root.compile = function (schema) {

        return Cast.schema(schema);
    };

    root.assert = function (value, schema, message) {

        message = (message ? message + ' ' : '');
        var error = root.validate(value, schema).error;
        if (error) {
            throw new Error(message + error.annotate());
        }
    };

    return root;
};


module.exports = internals.root();
